### PR TITLE
feat: add preloadInWorker API

### PR DESCRIPTION
**Pull request by hamst**
#### Description of Change
Add `preloadInWorker` for `webPreferences` in `BrowserWindow` to define a script to be loaded first before any other scripts in worker.

cc @codebytere 

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Added preloadInWorker to define a script to be loaded first before any other scripts in worker.

### CODE REVIEW

File: lib/worker/init.ts [hunk 1]
```diff
@@ -12,24 +12,61 @@ require('../common/reset-search-paths');
...
+
+const parseOption = function (name: string) {
+  return hasSwitch(name) ? getSwitchValue(name) : null;
+};
+
+const nodeIntegration = hasSwitch('node-integration-in-worker');
+const preloadScript = parseOption('preload-in-worker');
+
+if (nodeIntegration) {
+  // Export node bindings to global.
```
Comment (zcbenz) [inline lib/worker/init.ts:26]: This file shares lots of code with `lib/renderer/init.ts`, can you move the shared code into a utility function? Like what `windowSetup` and `webViewInit` do.

**Review by zcbenz — CHANGES_REQUESTED** [review]
From the document it is unclear how this option works with `contextIsolation`, can you make it clear and add related tests?

And I think this option should either respect the `contextIsolation` option of the window, or just always run the preload script in isolated context. /cc @electron/wg-api

**Comment by hamst** [comment]
@zcbenz It seems that contextIsolation is not supported in WebWorkers, so adding preloadInWorker doesn't affect it in any ways. I will add some details in documents, as far as having tests with contextIsolation, will they be useful?

**Comment by zcbenz** [comment]
The `contextIsolation` controls whether preload scripts should run in the web page's context or in a separate context, currently it is not related to WebWorker because WebWorker does not have preload scripts.

But since this PR is adding `preloadInWorker`, we have to define how `contextIsolation` affects preload scripts in WebWorker. We can either:
1. Make `preloadInWorker` respect the `contextIsolation` option that, `contextIsolation` should be able to control whether preload scripts should run in the worker script's context or in a separate context ;
2. Or ignore the  `contextIsolation` option and always run the preload scripts in separate context in WebWorker.

From current implementation, the preload scripts specified by `preloadInWorker` are running in the same context of the worker script, which is against the "secure by default" principle that we have in Electron's APIs.

**Review by MarshallOfSound — CHANGES_REQUESTED** [review]
Without solving the ctx isolation issue:

> This script will always have access to node APIs no matter whether node integration in worker is turned on or off.

is fundamentally a massive security vulnerability that we can't / won't land into this project.  Putting a hard block on this pending a proposal for how to deal with ctx isolation

**Comment by hamst** [comment]
Right now we have an option `nodeIntegrationInWorker` but don't support `contextIsolation` in worker. Being enabled `nodeIntegrationInWorker` allows access to powerful API without any limitation for all web worker scripts loaded in webpage, and there is no way to limit this. Very unlikely that this option without `contextIsolation` support follows "secure by default" principle, and that's what we have today. So `contextIsolation` for worker must be supported no matter if we have `preloadInWorker` or not.
Option `preloadInWorker` makes things way better: you define the script with all access to API, and you can make some API visible for all workers, but it requires additional steps and unlikely could be done unintentionally. By default, nothing will be exposed. That's a big difference from security prospective.
So getting `contextIsolation` support for workers should be a separate task which must be done to be "secure by defaults" regardless `preloadInWorker`. As `contextIsolation` heavily relies on contextBridge module, some work needs to be done to relax the requirement to avoid using electron modules in workers https://www.electronjs.org/docs/tutorial/multithreading#available-apis and make the implementation thread-safe.

**Comment by zcbenz** [comment]
@hamst Thanks for explaining, I see your point now. I'm good with this API, since it is an improvement over `nodeIntegrationInWorker` on security, before `contextIsolation` is supported in worker.

**Comment by hamst** [comment]
Would be good to link the issue #28620 to this PR.

**Pull request closed without merging**